system/lib/wasmfs/file.cpp

```diff
@@ -40,33 +40,180 @@ void DataFile::Handle::preloadFromJS(int index) {
+  auto& dcache = getDir()->dcache;
+  auto [_, inserted] = dcache.insert({name, {DCacheKind::Normal, child}});
+  assert(inserted && "inserted child already existed!");
+  assert(child->locked().getParent() == nullptr);
+  child->locked().setParent(getDir());
+}
+
+std::shared_ptr<File> Directory::Handle::getChild(const std::string& name) {
+  // Unlinked directories must be empty, without even "." or ".."
+  if (!getParent()) {
+    return nullptr;
+  }
+  if (name == ".") {
+    return file;
+  }
+  if (name == "..") {
+    return getParent();
+  }
+  // Check whether the cache already contains this child.
+  auto& dcache = getDir()->dcache;
+  if (auto it = dcache.find(name); it != dcache.end()) {
+    return it->second.file;
+  }
+  // Otherwise check whether the backend knows about the child.
+  auto child = getDir()->getChild(name);
+  if (!child) {
+    return nullptr;
+  }
+  finishInsertingChild(name, child);
+  return child;
+}
+
+bool Directory::Handle::mountChild(const std::string& name,
+                                   std::shared_ptr<File> child) {
+  assert(child);
+  // Cannot insert into an unlinked directory.
+  if (!getParent()) {
+    return false;
+  }
+  // Insert the entry into the cache and set its parent.
+  auto [_, inserted] =
+    getDir()->dcache.insert({name, {DCacheKind::Mount, child}});
+  assert(inserted && "mountChild called when file already existed!");
+  assert(child->locked().getParent() == nullptr);
+  child->locked().setParent(getDir());
+  return true;
+}
+
+std::shared_ptr<DataFile>
+Directory::Handle::insertDataFile(const std::string& name, mode_t mode) {
+  // Cannot insert into an unlinked directory.
+  if (!getParent()) {
+    return nullptr;
+  }
+  auto child = getDir()->insertDataFile(name, mode);
+  if (!child) {
+    return nullptr;
+  }
+  finishInsertingChild(name, child);
+  return child;
+}
+
+std::shared_ptr<Directory>
+Directory::Handle::insertDirectory(const std::string& name, mode_t mode) {
+  // Cannot insert into an unlinked directory.
+  if (!getParent()) {
+    return nullptr;
+  }
+  auto child = getDir()->insertDirectory(name, mode);
+  if (!child) {
+    return nullptr;
+  }
+  finishInsertingChild(name, child);
+  return child;
+}
+
+std::shared_ptr<Symlink>
+Directory::Handle::insertSymlink(const std::string& name,
+                                 const std::string& target) {
+  // Cannot insert into an unlinked directory.
+  if (!getParent()) {
+    return nullptr;
+  }
+  auto child = getDir()->insertSymlink(name, target);
+  if (!child) {
+    return nullptr;
+  }
+  finishInsertingChild(name, child);
+  return child;
+}
+
+bool Directory::Handle::insertMove(const std::string& name,
+                                   std::shared_ptr<File> file) {
+  // Cannot insert into an unlinked directory.
+  if (!getParent()) {
+    return false;
+  }
+  // Look up the file in its old parent's cache.
+  auto& oldCache = file->locked().getParent()->dcache;
```

Comment:
This locks `file`, and we hold a lock onto this directory (since we are a handle to it), but don't we need to take a lock on the old parent directory (if it is not the same as this directory) before we modify its dcache?

Reply:
No, because we can depend on this only being called by `__syscall_renameat`, which already holds the locks on both directories. It's kind of a code smell to have a dependency like that, working around it would mean bigger changes.

In the future we might consider replacing `insertMove` with a `Backend::move` that does not privilege either the source or destination directories. We might then consider passing `Directory::Handle`s as the arguments to prove that the directories have both been locked.

Reply:
I see, makes sense. I think a comment and a TODO are enough for now here.

Reply:
Sounds good. I added a TODO in one of the latest commits.